### PR TITLE
jwa: Improve how JWA exposes errors

### PR DESCRIPTION
**Pull request by elenzio9**
Currently, JWA doesn't properly expose the error state when errors occur in the underlying resources that the app is showing. Our goal it to further improve the way the app exposes errors to better inform users what's going on. So, in this PR:

#### Backend

1.  Extend the [process_status](https://github.com/kubeflow/kubeflow/blob/f2f223fc8781bfd0f6a0a59b6539ca5f229f28fe/components/crud-web-apps/jupyter/backend/apps/common/status.py#L9) function to consider the following cases:
     - Check the `.status.containerState`.
     -  Check the `.status.conditions`, since they have the one-liner `reason` and a `message`.
     -  If none of the above exist, it will use the `Events` emitted for the notebook.
     - In case it deduces the status from an `Event` and it's not available anymore, it will use a generic message.
2.  Add _a 10 second delay_ to the backend logic, where we display a spinner and a generic message to prevent a   warning icon from appearing just after a notebook is being initialized.
3.  Extend the [getNotebook](https://github.com/kubeflow/kubeflow/blob/f2f223fc8781bfd0f6a0a59b6539ca5f229f28fe/components/crud-web-apps/jupyter/backend/apps/common/routes/get.py#L59-L62) request to also include the _processed status_ information in the Notebook details page.

#### Frontend

1. Add an admonition in the details page of each Notebook with a detailed message on the current status.
2. Don't show the popup when a notebook is being stopped.

<details>
  <summary>Click here to show screenshots of how it looks like</summary>

image

image

image
</details>

### CODE REVIEW

**Comment by kimwnasptd** [comment]
Thanks for this effort @elenzio9! I'll also create a more generalized issue for this effort, since we'll also need to set some future goals for how the Notebooks Controller exposes the status right now.

Regarding this PR, could you also create some Cypress tests for the different cases of the status we expect? This way we'll be sure this expectation will be tested in future PRs

**Comment by elenzio9** [comment]
> Regarding this PR, could you also create some Cypress tests for the different cases of the status we expect? This way we'll be sure this expectation will be tested in future PRs

After looking at the Cypress tests I found this [one](https://github.com/kubeflow/kubeflow/blob/80539fca7fcffaacf6aaf911b767513f6510d705/components/crud-web-apps/jupyter/frontend/cypress/e2e/main-page.cy.ts#L55), which checks the status icon considering [any type of status](https://github.com/kubeflow/kubeflow/blob/80539fca7fcffaacf6aaf911b767513f6510d705/components/crud-web-apps/common/frontend/kubeflow-common-lib/projects/kubeflow/src/lib/resource-table/status/types.ts#L1-L10). So I think that covers us at least for now.

Once the frontend receives the raw YAML of each notebook (rather than the _processed status_ sent by the backend), we can go ahead and create more tests for the different cases of the status we expect, as then the preprocessing part will happen in the frontend. However, this requires improvements to be made on the Notebook Controller side in order to always expose the status in the Notebook's `.status.conditions`.

@kimwnasptd Wdyt?

**Review by tasos-ale — CHANGES_REQUESTED** [review]
Thanks for the PR @elenzio9. Please take a look at my comments.

**Comment by elenzio9** [comment]
> Thanks for the PR @elenzio9. Please take a look at my comments.

@tasos-ale Thanks for you comments! I just force pushed!

File: components/crud-web-apps/common/frontend/kubeflow-common-lib/projects/kubeflow/src/lib/status-icon/status-icon.component.scss [hunk 1]
```diff
@@ -1,15 +1,25 @@
 .small-padding-right {
   padding-right: 8px;
```
Comment (tasos-ale) [inline components/crud-web-apps/common/frontend/kubeflow-common-lib/projects/kubeflow/src/lib/status-icon/status-icon.component.scss:2]: @elenzio9 same thing I mention [here](https://github.com/kubeflow/kubeflow/pull/6952#discussion_r1132106601) also applies in this padding

**Review by tasos-ale — CHANGES_REQUESTED** [review]

File: components/crud-web-apps/common/frontend/kubeflow-common-lib/projects/kubeflow/src/lib/resource-table/resource-table.module.ts [hunk 2]
```diff
@@ -67,10 +67,10 @@ import { RouterModule } from '@angular/router';
     MatSelectModule,
     HelpPopoverModule,
     RouterModule,
+    StatusModule,
   ],
   declarations: [
     ResourceTableComponent,
-    StatusComponent,
     ActionComponent,
```
[the remark this answers is not in the document]
Comment (elenzio9) [reply components/crud-web-apps/common/frontend/kubeflow-common-lib/projects/kubeflow/src/lib/resource-table/resource-table.module.ts]: Yes, right!

File: components/crud-web-apps/common/frontend/kubeflow-common-lib/projects/kubeflow/src/lib/status-icon/status-icon.component.scss (hunk 1 above)
Comment (elenzio9) [reply components/crud-web-apps/common/frontend/kubeflow-common-lib/projects/kubeflow/src/lib/status-icon/status-icon.component.scss]: Done! Just pushed the changes!

**Comment by tasos-ale** [comment]
/lgtm
cc @kimwnasptd

File: components/crud-web-apps/jupyter/backend/apps/common/status.py [hunk 3]
```diff
@@ -8,54 +8,75 @@
...
-        if readyReplicas == 0:
-            return status.create_status(
-                status.STATUS_PHASE.STOPPED,
-                "No Pods are currently running for this Notebook Server.",
-            )
-        else:
-            return status.create_status(
-                status.STATUS_PHASE.TERMINATING, "Notebook Server is stopping."
-            )
+    creationTimestamp = notebook["metadata"]["creationTimestamp"]
```
Comment (kimwnasptd) [inline components/crud-web-apps/jupyter/backend/apps/common/status.py:14]: Let's rename this var to `creation_timestamp`. Python is using snake_case for var names 
https://peps.python.org/pep-0008/#descriptive-naming-styles

File: components/crud-web-apps/jupyter/backend/apps/common/status.py [hunk 4]
```diff
@@ -8,54 +8,75 @@
 
 def process_status(notebook):
     """
-    Return status and reason. Status may be [running|waiting|warning|error]
+    Return status and reason. Status may be [running|waiting|warning]
     """
-    # Check if the Notebook is stopped
     readyReplicas = notebook.get("status", {}).get("readyReplicas", 0)
```
Comment (kimwnasptd) [inline components/crud-web-apps/jupyter/backend/apps/common/status.py:13]: while not touched by this PR, let's also change this one to snake_case

File: components/crud-web-apps/jupyter/backend/apps/common/status.py [hunk 5]
```diff
@@ -8,54 +8,75 @@
...
+    while delta.total_seconds() <= 10:
+        status_val, reason = status.STATUS_PHASE.WAITING, "Waiting for StatefulSet to create the underlying Pod."
+
+        # Check if the Notebook is stopped or deleted
+        status_phase, status_message = check_stopped_deleted_nb(notebook)
+        if status_phase is not None:
+            status_val, reason = status_phase, status_message
+
+        return status.create_status(status_val, reason)
+    else:
```
Comment (kimwnasptd) [inline components/crud-web-apps/jupyter/backend/apps/common/status.py:31]: lets remove this `else`, since the above is an `if` with a `return`

**Review by kimwnasptd — COMMENTED** [review]
Did a first pass and added some comments as I took a look at the backend changes. Will add more in the next days

**Comment by elenzio9** [comment]
> Did a first pass and added some comments as I took a look at the backend changes. Will add more in the next days

@kimwnasptd Thanks for the comments! Just pushed the changes in fixup commits.

**Pull request merged**